### PR TITLE
Patch langchain async APIs for autologging

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/12107?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12107/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12107
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Patch `ainvoke`, `abatch`, `astream` methods for runnables.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/langchain/__init__.py [hunk 1]
```diff
@@ -967,10 +967,6 @@ def autolog(
     log_model_signatures=False,
     log_models=False,
     log_datasets=False,
-    # TODO: log_inputs_outputs was originally defaulted to True in the production version of
-    # the LangChain autologging, as it is a common use case to log input/output table for
-    # evaluation. Once tracing is fully launched, this should be supported by the tracer
-    # but we should design it to be compatible with the existing UJ.
```
Comment (serena-ruan) [inline mlflow/langchain/__init__.py:973]: We don't need to fix this anymore in my opinion, we should probably drop this in the future version. Logging inputs & outputs is error-prone when converting to dataframe.

File: mlflow/langchain/_langchain_autolog.py [hunk 2]
```diff
@@ -253,7 +253,7 @@ def _inject_callbacks(original_callbacks, new_callbacks):
 
     def _is_new_callback_already_in_original_callbacks(new_callback, original_callbacks):
         return any(
-            isinstance(new_callback, type(original_callback))
+            isinstance(original_callback, type(new_callback))
```
Comment (serena-ruan) [inline mlflow/langchain/_langchain_autolog.py:256]: This seems the same as old ways, but should we check if the new_callback instances already exist in original_callbacks? cc @dbczumar

File: mlflow/langchain/__init__.py [hunk 3]
```diff
@@ -929,7 +929,7 @@ def _patch_runnable_cls(cls):
     For classes that are subclasses of Runnable, we patch the `invoke`, `batch`, and `stream`
     methods for autologging.
     """
-    for func_name in ["invoke", "batch", "stream"]:
+    for func_name in ["invoke", "batch", "stream", "ainvoke", "abatch", "astream"]:
```
Comment (harupy) [inline mlflow/langchain/__init__.py:932]: `patched_inference` is non async function, right?

Comment (serena-ruan) [reply mlflow/langchain/__init__.py:932]: Ah it's not, but it works 🤔

Comment (harupy) [reply mlflow/langchain/__init__.py:932]: I think what's happening is this:

```
import asyncio


async def original():
    await asyncio.sleep(1)
    return 0


def patch():
    return original()


async def main():
    print(await patch())


if __name__ == "__main__":
    asyncio.run(main())

```

**Pull request merged**